pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+	version := version(next)
+	if version == invalidVersion {
+		return errInvalidVersion
+	}
+
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			slog.Error("RemoveAt may not be called with version lower than existing already", logfields.NewVersion, version, logfields.OldVersion, node.versions.first, logfields.Stacktrace, stacktrace, logfields.LogSubsys, "containers/versioned")
+			return errStaleVersion
+		}
+
+		if node.versions.contains(version) {
+			// Truncate the validity of this node to end at 'version'.
+			// After this readers with 'version' and above no longer see this value,
+			// while readers with versions before 'version' still see this.
+			node.versions.past.store(version)
+			break
+		}
+	}
+	return nil
+}
+
+// RemoveBefore removes all values whose validity ends before 'keepVersion'.
+// Caller must coordinate for mutual exclusion.
+func (v *Value[T]) RemoveBefore(keepVersion KeepVersion) {
+	version := version(keepVersion)
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.load() <= version {
+		// This node is no longer visible for readers with 'version' and above,
+		// so this can be safely removed.
+		node = node.next.Load()
+	}
+	v.head.Store(node)
+}
+
+// At returns value of type 'T' valid for the given version, or an empty value if none is found.
+func (v *Value[T]) At(hold *VersionHold) T {
+	version := invalidVersion
+	if hold != nil {
+		version = hold.version
+	}
+	if version == invalidVersion {
+		stacktrace := hclog.Stacktrace()
+		slog.Error("At: Invalid VersionHold finds nothing", logfields.Stacktrace, stacktrace, logfields.LogSubsys, "containers/versioned")
+	}
+
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if node.versions.contains(version) {
+			return node.value
+		}
+	}
+	var empty T
+	return empty
+}
+
+// Versioned is a pair of a version and any type T
+type Versioned[T any] struct {
+	version version
+	value   T
+}
+
+type VersionedSlice[T any] []Versioned[T]
+
+// Append appends a pair of 'nextVersion' and 'value' to VersionedSlice 's', returning updated
+// 's'. Needed to keep members private.
+// Should only be called with monotonically increasing 'nextVersion's, so that the slice remains
+// sorted by version in ascending order.
+func (s VersionedSlice[T]) Append(nextVersion NextVersion, value T) VersionedSlice[T] {
+	return append(s, Versioned[T]{
+		version: version(nextVersion),
+		value:   value,
+	})
+}
+
+// ForEachBefore iterates the elements in VersionedSlice 's', calling function 'f(T)' with each
+// value associated with versions earlier than 'keepVersion'.
+// The slice is assumed to be sorted by version in ascending order.
+// Returns the number of calls made.
+func (s VersionedSlice[T]) ForEachBefore(keepVersion KeepVersion, f func(T)) (n int) {
```

Comment:
With the new range func in Go 1.23, this should be an iterator instead, so that folks can do

```go
for t := range vslice.ForEachBefore(ver) {
   // whatever they want with t
}
```

or even

```go
for v, t := range vslice.ForEachBefore(ver) {
   // whatever they want with t and v
}
```

Reply:
Nice! Will do when we get to Go 1.23!

Reply:
We now do have v1.23 in tree. Would it make sense to adapt this now or are you planning to do it in a followup?

Reply:
Done!